### PR TITLE
Feature/kyv 560 create accounting from refund accounting data

### PR DESCRIPTION
**Pull request by laukkanenj**

### CODE REVIEW

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/api/data/accounting/AccountingSlipRowDto.java [hunk 1]
```diff
@@ -51,6 +51,9 @@ public class AccountingSlipRowDto {
     @JacksonXmlProperty(localName = "FunctionalArea")
     private String functionalArea;
 
+    // "refund" or "order"
+    private String rowType;
```
Comment (severikupari1) [inline orderapi/src/main/java/fi/hel/verkkokauppa/order/api/data/accounting/AccountingSlipRowDto.java:55]: Could this be enum?

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/api/data/accounting/RefundItemAccountingDto.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.util.Objects;
+
+/**
+ * Yritys
```
Comment (severikupari1) [inline orderapi/src/main/java/fi/hel/verkkokauppa/order/api/data/accounting/RefundItemAccountingDto.java:11]: Could these be moved just before variable?

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/ServiceUtils.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package fi.hel.verkkokauppa.order.service;
+
+import org.springframework.stereotype.Service;
+
+import java.util.HashSet;
+import java.util.Set;
+
+@Service
+public class ServiceUtils {
```
Comment (severikupari1) [inline orderapi/src/main/java/fi/hel/verkkokauppa/order/service/ServiceUtils.java:9]: Could this be moved to common library?

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/AccountingExportService.java [hunk 4]
```diff
@@ -157,4 +167,80 @@ private void markAsExported(AccountingExportDataDto exportData) {
...
+        // build set of all balanceProfitCenters
+        Set<String> balanceProfitCenters = serviceUtils.combineKeySets(
+                ordersByBalanceProfitCenter.keySet(),
+                refundsByBalanceProfitCenter.keySet(),
+                String.class
+        );
+
+        // collect all rows by balance profit center
+        List<AccountingSlipRowDto> newRows = new ArrayList<>();
+//        for (Map.Entry<String, List<AccountingSlipRowDto>> orderedAccountings : ordersByBalanceProfitCenter.entrySet()) {
```
Comment (severikupari1) [inline orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/AccountingExportService.java:194]: Remove this commented out code?

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/AccountingExportService.java [hunk 5]
```diff
@@ -157,4 +167,80 @@ private void markAsExported(AccountingExportDataDto exportData) {
...
+        return accountingSlipWithRows;
+    }
+
+    private void splitAccountingSlipRowsToOrdersAndRefunds(
+            List<AccountingSlipRowDto> allRows,
+            List<AccountingSlipRowDto> orderRows,
+            List<AccountingSlipRowDto> refundRows) {
+
+        for (AccountingSlipRowDto row : allRows) {
+            if (row.getRowType().contains("refund")) {
```
Comment (severikupari1) [inline orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/AccountingExportService.java:237]: This should be equals and use enum in here.

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/RefundAccountingService.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+public class RefundAccountingService {
+
+    private Logger log = LoggerFactory.getLogger(RefundAccountingService.class);
```
Comment (severikupari1) [inline orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/RefundAccountingService.java:16]: This logger can be done with class annotation `@Slf4j`

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/RefundItemAccountingService.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    private Logger log = LoggerFactory.getLogger(RefundItemAccountingService.class);
+
+    @Autowired
+    private RefundItemAccountingRepository refundItemAccountingRepository;
+
+    public List<RefundItemAccounting> getRefundItemAccountings(String refundId) {
+        List<RefundItemAccounting> accountings = refundItemAccountingRepository.findByRefundId(refundId);
+
+        if (accountings.size() > 0)
+            return accountings;
```
Comment (severikupari1) [inline orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/RefundItemAccountingService.java:25]: Add braces to if

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/refund/RefundService.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+import org.springframework.stereotype.Component;
+
+import java.time.LocalDate;
+import java.util.Optional;
+
+
+@Component
+public class RefundService {
+
+    private Logger log = LoggerFactory.getLogger(RefundService.class);
```
Comment (severikupari1) [inline orderapi/src/main/java/fi/hel/verkkokauppa/order/service/refund/RefundService.java:18]: This logger can be done with class annotation @Slf4j

**Review by severikupari1 — CHANGES_REQUESTED** [review]
Some minor changes

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/model/refund/Refund.java [hunk 9]
```diff
@@ -10,72 +10,84 @@
...
+    @Field(type = FieldType.Text)
+    String priceTotal;
+
+    @Field(type = FieldType.Date, format = DateFormat.date)
+    LocalDate accounted;
+
+    @Field(type = FieldType.Text)
+    String accountingStatus;
+
+    public Refund() {
```
Comment (severikupari1) [inline orderapi/src/main/java/fi/hel/verkkokauppa/order/model/refund/Refund.java:70]: FYI, you can use lombok `@NoArgsConstructor` to do this too.

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/model/refund/RefundAccountingStatus.java [hunk 10]
```diff
@@ -0,0 +1,6 @@
+package fi.hel.verkkokauppa.order.model.refund;
+
+public class RefundAccountingStatus {
+    public static String CREATED = "created";
```
Comment (severikupari1) [inline orderapi/src/main/java/fi/hel/verkkokauppa/order/model/refund/RefundAccountingStatus.java:4]: Just asking can this be enum or not?

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/api/data/accounting/AccountingSlipRowDto.java (hunk 1 above)
Comment (laukkanenj) [reply orderapi/src/main/java/fi/hel/verkkokauppa/order/api/data/accounting/AccountingSlipRowDto.java]: yes, done

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/api/data/accounting/RefundItemAccountingDto.java (hunk 2 above)
Comment (laukkanenj) [reply orderapi/src/main/java/fi/hel/verkkokauppa/order/api/data/accounting/RefundItemAccountingDto.java]: done

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/model/accounting/RefundAccounting.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+import lombok.Data;
+import org.springframework.data.annotation.Id;
+import org.springframework.data.elasticsearch.annotations.DateFormat;
+import org.springframework.data.elasticsearch.annotations.Document;
+import org.springframework.data.elasticsearch.annotations.Field;
+import org.springframework.data.elasticsearch.annotations.FieldType;
+
+import java.time.LocalDateTime;
+
+@Document(indexName = "refundaccountings")
```
[the remark this answers is not in the document]
Comment (laukkanenj) [reply orderapi/src/main/java/fi/hel/verkkokauppa/order/model/accounting/RefundAccounting.java]: can be. Done and also renamed the refunditemaccountings to refund_item_accountings

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/ServiceUtils.java (hunk 3 above)
Comment (laukkanenj) [reply orderapi/src/main/java/fi/hel/verkkokauppa/order/service/ServiceUtils.java]: yes, done

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/AccountingExportService.java (hunk 4 above)
Comment (laukkanenj) [reply orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/AccountingExportService.java]: done

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/RefundAccountingService.java (hunk 6 above)
Comment (laukkanenj) [reply orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/RefundAccountingService.java]: done

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/RefundItemAccountingService.java (hunk 7 above)
Comment (laukkanenj) [reply orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/RefundItemAccountingService.java:25]: done

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/refund/RefundService.java (hunk 8 above)
Comment (laukkanenj) [reply orderapi/src/main/java/fi/hel/verkkokauppa/order/service/refund/RefundService.java]: done

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/model/refund/RefundAccountingStatus.java (hunk 10 above)
Comment (laukkanenj) [reply orderapi/src/main/java/fi/hel/verkkokauppa/order/model/refund/RefundAccountingStatus.java:4]: probably

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/AccountingSlipService.java [hunk 12]
```diff
@@ -272,6 +404,43 @@ private List<AccountingSlipRowDto> createAccountingSlipRowDtos(List<OrderItemAcc
...
+
+        for (RefundItemAccountingDto summedItemAccounting : summedItemAccountings) {
+            String accountingSlipRowId = UUIDGenerator.generateType3UUIDString(accountingSlipId, Integer.toString(rowNumber));
+
+            AccountingSlipRowDto accountingSlipRowDto = AccountingSlipRowDto.builder()
+                    .accountingSlipRowId(accountingSlipRowId)
+                    .accountingSlipId(accountingSlipId)
+                    .taxCode(summedItemAccounting.getVatCode())
+                    .amountInDocumentCurrency(formatSum(summedItemAccounting.getPriceGrossAsDouble()))
+                    .baseAmount("+" + formatSum(summedItemAccounting.getPriceNetAsDouble()))
```
Comment (severikupari1) [inline orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/AccountingSlipService.java:434]: Is that + correct when using refunds?

**Review by severikupari1 — COMMENTED** [review]
Good job, check that one question about using + in refund values

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/AccountingSlipService.java (hunk 12 above)
Comment (laukkanenj) [reply orderapi/src/main/java/fi/hel/verkkokauppa/order/service/accounting/AccountingSlipService.java:434]: for orders there is automatically - for sum rows and + added for balance rows (always + or - before numbers). For refunds I added + for sum rows and there is automatically - for balance rows. Jarkko approved the example xml.

**Review by severikupari1 — APPROVED** [review]

**Pull request merged**